pretyping/pretyping.ml

```diff
@@ -186,9 +186,23 @@ type inference_flags = {
 let frozen_holes (sigma, sigma') =
   (); fun ev -> Evar.Map.mem ev (Evd.undefined_map sigma)
 
+let rec advance sigma evk =
+  let evi = Evd.find sigma evk in
+  match evi.evar_body with
+  | Evar_empty -> Some evk
+  | Evar_defined v ->
+      if Option.default false (Store.get evi.evar_extra Evarutil.cleared) then
+        let (evk,_) = Term.destEvar v in
+        advance sigma evk
+      else
+        None
+
 let pending_holes (sigma, sigma') =
+  let add evk evi acc =
+    match advance sigma' evk with None -> acc | Some evk' -> Evar.Set.add evk' acc in
+  let initial = Evd.fold_undefined add sigma Evar.Set.empty in
```

Comment:
I think your proposal does not work. Are initial those undefined evars of Σ that are still undefined in Σ' and those undefined evars of Σ' which derive by restriction (e.g. `clear`) from an undefined evar in Σ.

As requested, I put a comment on frozen_holes and pending_holes, and actually merged them. PMP took time to make an efficiency improvement on testing the frozen state but I don't know how to preserve it because an evar ?x in Σ knows that it gave birth to an evar ?x' in Σ', but Σ' does not know that it derives from ?x. (So somehow, a "cleared" flag should be also on ?x' if I'm not mistaken.

I also moved advance in evarutil.ml.